compiler/rustc_lint/src/function_cast_as_integer.rs

```diff
@@ -0,0 +1,60 @@
+use rustc_hir as hir;
+use rustc_middle::ty;
+use rustc_session::{declare_lint, declare_lint_pass};
+use rustc_span::BytePos;
+
+use crate::lints::{FunctionCastsAsIntegerDiag, FunctionCastsAsIntegerSugg};
+use crate::{LateContext, LateLintPass};
+
+declare_lint! {
+    /// The `function_casts_as_integer` lint detects cases where a function item is cast
+    /// to an integer.
+    ///
+    /// ### Example
+    ///
+    /// ```rust
+    /// fn foo() {}
+    /// let x = foo as usize;
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// When casting a function item to an integer, it implicitly creates a
+    /// function pointer that will in turn be cast to an integer. By making
+    /// it explicit, it improves readability of the code and prevents bugs.
+    pub FUNCTION_CASTS_AS_INTEGER,
+    Warn,
+    "casting a function into an integer",
+}
+
+declare_lint_pass!(
+    /// Lint for casts of functions into integers.
+    FunctionCastsAsInteger => [FUNCTION_CASTS_AS_INTEGER]
+);
+
+impl<'tcx> LateLintPass<'tcx> for FunctionCastsAsInteger {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx hir::Expr<'tcx>) {
+        let hir::ExprKind::Cast(cast_from_expr, cast_to_expr) = expr.kind else { return };
+        let cast_to_ty = cx.typeck_results().expr_ty(expr);
+        // Casting to a function (pointer?), so all good.
+        if matches!(cast_to_ty.kind(), ty::FnDef(..) | ty::FnPtr(..) | ty::RawPtr(..)) {
+            return;
+        }
```

Comment:
It is probably better to check that the type which is being cast to is actually a integer, instead of checking if it's not something else.
```suggestion
        // Check if it's casting to a integer.
        if !matches!(cast_to_ty.kind(), ty::Int(..) | ty::Uint(..)) {
            return;
        }
```

Reply:
I disagree here. The point is to check if a function pointer is casted as something else than a pointer.

Reply:
The name of the lint and the diagnostic both only mentions integers. If we want to lint to things other than cast to integer it would be good to adapt both, otherwise it's going to be confusing to call a cast to integer when it's not the case.

But other than function pointer and integer, is it even possible to cast a function definition to something else?

Reply:
It's the only supported case for now. But in case it might change in the future...